Firestore/core/src/firebase/firestore/immutable/llrb_node.h

```diff
@@ -258,6 +288,63 @@ LlrbNode<K, V> LlrbNode<K, V>::InnerInsert(const K& key,
   return result;
 }
 
+template <typename K, typename V>
+template <typename Comparator>
+LlrbNode<K, V> LlrbNode<K, V>::erase(const K& key,
+                                     const Comparator& comparator) const {
+  LlrbNode root = InnerErase(key, comparator);
+  root.FixRootColor();
+  return root;
+}
+
+template <typename K, typename V>
+template <typename Comparator>
+LlrbNode<K, V> LlrbNode<K, V>::InnerErase(const K& key,
+                                          const Comparator& comparator) const {
+  if (empty()) {
+    // Empty node already frozen
+    return LlrbNode{};
+  }
+
+  LlrbNode n = Clone();
+
+  bool descending = comparator(key, n.key());
+  if (descending) {
+    if (!n.left().empty() && !n.left().red() && !n.left().left().red()) {
+      n.MoveRedLeft();
+    }
+    n.set_left(n.left().InnerErase(key, comparator));
+
+  } else {
+    if (n.left().red()) {
+      n.RotateRight();
+    }
+
+    if (!n.right().empty() && !n.right().red() && !n.right().left().red()) {
+      n.MoveRedRight();
+    }
+
+    if (util::Compare(key, n.key(), comparator) ==
```

Comment:
Optional: I think it would be nice to split this somewhat long function and maybe reduce nesting a little bit. One thing that comes to mind: lines 319-325 apply to both equal and greater cases, would it be possible to move them to a helper method or a local lambda? Then the `== Same` case could be moved out of this branch into a top-level branch:
```cpp
auto order = util::Compare(key, n.key(), comparator);
if (order == util::ComparisonResult::Descending) {
  // ... same as previous if (descending) branch ...
  n.set_left(n.left().InnerErase(key, comparator));

} else if (order == util::ComparisonResult::Ascending) {
  FixUpRightSubtree(); // Strawman name; can be a lambda
  n.set_right(n.right().InnerErase(key, comparator)); // This is now clearly symmetrical with the descending case

} else { // must be equal
  FixUpRightSubtree(); // Lines 313-315 could symmetrically be FixUpLeftSubtree
  // ...equivalent to lines 329-339
}
```